azblob/chunkwriting.go

```diff
@@ -0,0 +1,214 @@
+
+	// num is the current chunk we are on.
+	num int32
+	// ch is used to pass the next chunk of data from our reader to one of the writers.
+	ch chan *chunk
+	// errCh is used to hold the first error from our concurrent writers.
+	errCh chan error
+	// wg provides a count of how many writers we are waiting to finish.
+	wg sync.WaitGroup
+	// buffers provides a pool of chunks that can be reused.
+	buffers sync.Pool
+
+	// result holds the final result from blob storage after we have submitted all chunks.
+	result *BlockBlobCommitBlockListResponse
+}
+
+// getErr returns an error by priority. First, if a function set an error, it returns that error. Next, if the Context has an error
+// it returns that error. Otherwise it is nil. getErr supports only returning an error once per copier.
+func (c *copier) getErr() error {
+	select {
+	case err := <-c.errCh:
+		return err
+	default:
+	}
+	return c.ctx.Err()
+}
+
+// sendChunk reads data from out internal reader, creates a chunk, and sends it to be written via a channel.
+// sendChunk returns io.EOF when the reader returns an io.EOF or io.ErrUnexpectedEOF.
+func (c *copier) sendChunk() error {
+	if err := c.getErr(); err != nil {
+		return err
+	}
+
+	chunk := c.buffers.Get().(*chunk)
+	n, err := io.ReadFull(c.reader, chunk.payload)
+	chunk.payload = chunk.payload[0:n]
+	switch {
+	case err == nil && n == 0:
+		return nil
+	case err == nil:
+		chunk.num = c.num
+		c.num++
+		c.ch <- chunk
+		return nil
+	case err != nil && (err == io.EOF || err == io.ErrUnexpectedEOF) && n == 0:
+		return io.EOF
+	}
+
+	if err == io.EOF || err == io.ErrUnexpectedEOF {
+		chunk.num = c.num
+		c.num++
+		c.ch <- chunk
+		return io.EOF
+	}
+	if err := c.getErr(); err != nil {
+		return err
+	}
+	return err
+}
+
+// writer writes chunks sent on a channel.
+func (c *copier) writer() {
+	defer c.wg.Done()
+
+	for chunk := range c.ch {
+		if err := c.write(chunk); err != nil {
+			if !errors.Is(err, context.Canceled) {
+				select {
+				case c.errCh <- err:
+					c.cancel()
+				default:
+				}
+				return
+			}
+		}
+	}
+}
+
+// write uploads a chunk to blob storage.
+func (c *copier) write(chunk *chunk) error {
+	defer c.buffers.Put(chunk)
+
+	if err := c.ctx.Err(); err != nil {
+		return err
+	}
+
+	blockID := newUuidBlockID(c.prefix).WithBlockNumber(uint32(chunk.num)).ToBase64()
+	_, err := c.to.StageBlock(c.ctx, blockID, bytes.NewReader(chunk.payload), LeaseAccessConditions{}, nil)
+	if err != nil {
+		return fmt.Errorf("write error: %w", err)
+	}
+	return nil
+}
+
+// close commits our blocks to blob storage and closes our writer.
+func (c *copier) close() error {
+	close(c.ch)
+	c.wg.Wait()
+
+	if err := c.getErr(); err != nil {
+		return err
+	}
+
+	blockID := newUuidBlockID(c.prefix)
+	var blockIDs []string
+	if c.num > 0 {
+		blockIDs = make([]string, 0, c.num-1)
+		for i := 0; i < int(c.num); i++ {
+			blockIDs = append(blockIDs, blockID.WithBlockNumber(uint32(i)).ToBase64())
+		}
+	}
+
```

Comment:
_typically_ in azcopy, we just use raw GUIDs for our block IDs, so I'd recommend following a similar methodology (in case a user suddenly decides to try to use azcopy to overwrite it or something)

But that's down to how @zezha-msft and @JohnRusk feel about it. I'm not _opposed_ to this the way it is, I'm just mildly antsy about it.

Reply:
I assume you mean just use blockID?  

I'm pretty sure I just copied this method from the original code.  I'm good with whatever you all decide is best.  

@adreed-msft : could you and ze let me know how you would like it and I'll be glad to make thechange.

Reply:
Good point @adreed-msft, yes I agree that we should use Raw GUIDs to be consistent.

Reply:
Do you want that in this CL or another CL?

Reply:
this PR would be nice.

Reply:
Me and @element-of-surprise spoke offline.

Turns out this wouldn't break next to azcopy, because the SDK already pads the block ID at 64 bytes.

Though, I do agree with the "resumability" aspect of this, as well as switching to google/uuid quite a bit.